### ISSUE TITLE
How do I build the example?

### ISSUE THREAD

**Issue by vi** [zargony/rust-fuse#46]
`cargo build` produce empty `target/debug/examples/` directory.

Manual compilation also fails:

```
$  rustc    -L target/debug  examples/hello.rs  
examples/hello.rs:3:1: 3:19 error: found possibly newer version of crate `log` which `fuse` depends on
examples/hello.rs:3 extern crate fuse;
                    ^~~~~~~~~~~~~~~~~~
examples/hello.rs:3:1: 3:19 note: perhaps this crate needs to be recompiled?
examples/hello.rs:3 extern crate fuse;
                    ^~~~~~~~~~~~~~~~~~
examples/hello.rs:3:19: 3:19 note: crate `log` path #1: /home/vi/home/rust/prefix/lib/rustlib/i686-unknown-linux-gnu/lib/liblog-4e7c5e5c.rlib
examples/hello.rs:3:19: 3:19 note: crate `log` path #2: /home/vi/home/rust/prefix/lib/rustlib/i686-unknown-linux-gnu/lib/liblog-4e7c5e5c.so
examples/hello.rs:3:19: 3:19 note: crate `fuse` path #1: /home/vi/home/rust/rust-fuse/target/debug/libfuse-c61beed3b50079b8.rlib
error: aborting due to previous error
```

I think there should also be a `fusexmp` or `bindfs`-like example. This is both useful as a reference of Rust's filesytem features and as a basis for new FUSE programs.

**Comment by cuviper** [comment]
Cargo only builds examples with `cargo test`.  Add `--no-run` to build without running tests, and `-v` if you want to see the full commands for manual compilation.

As for additional examples, please feel free to contribute!  But I think if an example gets big enough to be genuinely useful, it will probably deserve its own project.  Perhaps we should list a few known projects in the README?  For instance, mine is [rust-git-fs](https://github.com/cuviper/rust-git-fs), which doesn't do a lot yet, but I've kept it maintained at least.

**Comment by vi** [comment]
Found the issue in manual compilation: `rustc    -L target/debug  -L target/debug/deps   examples/hello.rs` works.

I think the `fusexmp`-type example is fundamental enough to reside in library project itself and be maintained with the library.

**Comment by cuviper** [comment]
Looks like `fusexmp` is basically a pass-through?  Sure, I agree that's a good example to have.  You could probably implement that trivially over functions from the `libc` crate, but the result wouldn't be very rust-y.  Something implemented on `std::io` would be nicer, but requires more translation.

**Comment by vi** [comment]
There can be both, especially if `std::io` is not ready to implement everything `fusexmp` may need (such as extended attributes). But if there would be complete rust-fusexmp based on `std::io`, it would serve a great hub of Rust filesystem features (becase of it requires almost all filesystem features by design).

libc-based one can be done first (if that's simple).

**Comment by bolinfest** [comment]
I don't know whether this should be a separate issue, but I think including the following setup steps for OS X would be helpful:

1. Install `osxfuse` from https://osxfuse.github.io/.
2. Install `pkg-config`. This is very easy to install using Brew: `brew install pkg-config`. To verify it can find `osxfuse`, run `pkg-config --libs --cflags osxfuse`. You should get something like: `-D_FILE_OFFSET_BITS=64 -D_DARWIN_USE_64_BIT_INODE -I/usr/local/include/osxfuse/fuse -L/usr/local/lib -losxfuse -pthread -liconv`.
3. Create a directory where you want to mount the "hello" filesystem: `mkdir -p /tmp/helloexample`.
4. Run the example: `cargo run --example hello /tmp/helloexample/`.
5. Because the example runs in the foreground, in a new terminal, play around with your virtual filesystem!

```
$ ls -al /tmp/helloexample/
total 8
drwxr-xr-x   2 501   staff    0 Oct  8  2013 .
drwxrwxrwt  17 root  wheel  578 Jan 23 14:17 ..
-rw-r--r--   1 501   staff   13 Oct  8  2013 hello.txt
$ cat /tmp/helloexample/hello.txt
Hello World!
```

If you go back to the terminal where you launched the filesystem, it seems a little problematic once you kill the program with ctrl-C. Specifically, if you try to `ls` the directory afterwards:

```
^C
$ ls -al /tmp/helloexample/
ls: : Device not configured
$ rmdir /tmp/helloexample/
rmdir: /tmp/helloexample/: Resource busy
$ rm -rf /tmp/helloexample/
rm: /tmp/helloexample/: Device not configured
```

You get errors.

**Comment by bolinfest** [comment]
The problem with ctrl-C seems to be noted in https://github.com/zargony/rust-fuse/issues/58.

**Comment by vi** [comment]
The problem was that I didn't knew about `cargo run --example`.

I started doing #53, although it's in early stage of development.

**Comment by nicferrier** [comment]
I'm getting this:

```
cargo build
   Compiling memtarfs v0.1.0 (/home/nicferrier/rustwork/memtarfs)
error[E0308]: mismatched types
  --> src/main.rs:13:12
   |
13 |     atime: UNIX_EPOCH,                                  // 1970-01-01 00:00:00
   |            ^^^^^^^^^^ expected struct `time::Timespec`, found struct `std::time::SystemTime`
   |
   = note: expected type `time::Timespec`
              found type `std::time::SystemTime`
```

which seems to be related to hello.rs using the wrong value for it's types in the struct:

```rust

const HELLO_DIR_ATTR: FileAttr = FileAttr {
    ino: 1,
    size: 0,
    blocks: 0,
    atime: UNIX_EPOCH,                                  // 1970-01-01 00:00:00
    mtime: UNIX_EPOCH,
    ctime: UNIX_EPOCH,
    crtime: UNIX_EPOCH,
    kind: FileType::Directory,
    perm: 0o755,
    nlink: 2,
    uid: 501,
    gid: 20,
    rdev: 0,
    flags: 0,
};
```

But it can't be the wrong value since it compiles just fine inside the project with `cargo run`.

I made a another copy of hello.rs [here](https://github.com/nicferrier/rust-fuse-demo) to try to separate it.

Any folks on this project wanna tell me what I'm doing wrong?

**Comment by zargony** [comment]
rust-fuse 0.3.1 (the current release on crates.io) still uses `time::Timespec`. A change to `std::time` has been merged, but is still unreleased.

**Comment by nicferrier** [comment]
Thanks!

I'm still confused why the current version would compile and run the examples when I can't extract the same example, separate it into it's own repo and build it there... I guess because you haven't published this version to cargo yet? 

But thanks for pointing out the error; with another small change to the hello example I've got something that compiles! thank you!
